omegaconf/basecontainer.py

```diff
@@ -723,11 +728,16 @@ def prepand(full_key: str, parent_type: Any, cur_type: Any, key: Any) -> str:
             full_key = self._key()
 
         assert cur is not None
+        memo = {id(cur)}  # remember already visited nodes so as to detect cycles
         while cur._get_parent() is not None:
             cur = cur._get_parent()
+            if id(cur) in memo:
+                raise OmegaConfBaseException(
```

Comment:
I didn't really intend OmegaConfBaseException to be used directly.

We can add a user facing generic exception type `OmegaConfException(OmegaConfBaseException)`, or just use one of the built in exception types here (I can't think of any that fits well though).

Reply:
The generic builtin type is `RuntimeError` => I could just raise that if we don't really care about it inheriting from `OmegaConfBaseException` (another option could be `RecursionError`, which is thematically fit but technically wrong since there is no actual recursion).

Two other options would be:
- `OmegaConfException(OmegaConfBaseException)` as you suggested
- A more specific `ConfigDependencyError(OmegaConfBaseException)` which could be re-used in the future in any other situation where we detect a similar problem with the dependency graph between nodes.

I don't have a strong opinion on any of these, so let me know what you prefer.

Reply:
I guess with could introduce a `ConfigCycleDetectedException(OmegaConfBaseException)` in the spirit of your last suggestion.
seems generic enough and we may reuse it in other similar scenarios.

Reply:
Sounds good, done in 650e710